Tests/Unit/LoRaTools/TwinCollectionExtensionsTests.cs

```diff
@@ -0,0 +1,204 @@
+            Assert.False(tc.TryRead<string>("invalid", out _));
+        }
+
+        [Fact]
+        public void Null_Values_Are_Returned_As_Null()
+        {
+            var tc = new TwinCollection();
+            const string key = "test";
+            tc[key] = null;
+            var reader = new TwinCollectionReader(tc, this.logger);
+            Assert.Null(reader.SafeRead<string>(key));
+            Assert.False(reader.TryRead<string>(key, out _));
+        }
+
+        [Theory]
+        [InlineData(typeof(string), "test", "'test'")]
+        [InlineData(typeof(int),   int.MinValue)]
+        [InlineData(typeof(int),   int.MaxValue)]
+        [InlineData(typeof(bool),  true, "true")]
+        [InlineData(typeof(ulong), ulong.MaxValue)]
+        [InlineData(typeof(ushort),ushort.MaxValue)]
+        public void TryRead_Returns_True_When_Key_Exists_And_Correct_Value_Is_Returned(Type targetType, object someValue, string jsonValue = null)
+        {
+            const string key = "test";
+
+            var tr = CreateTwinCollectionReader(key, jsonValue ?? someValue);
+            var openMethod = typeof(TwinCollectionReader).GetMethod("TryRead");
+            var typedMethod = openMethod.MakeGenericMethod(targetType);
+            var parameters = new object[] { key, null };
+            Assert.True((bool)typedMethod.Invoke(tr, parameters));
+            Assert.Equal(someValue, parameters[^1]);
+        }
+
+        [Fact]
+        public void TryRead_Overflows_Are_Handled()
+        {
+            const string key = "test";
+            var tc = CreateTwinCollectionReader(key, ulong.MaxValue);
+            Assert.False(tc.TryRead<int>(key, out _));
+        }
+
+        [Fact]
+        public void TryRead_InvalidTypeConversions_Are_Handled()
+        {
+            const string key = "test";
+            var tc = CreateTwinCollectionReader(key, "true");
+            Assert.False(tc.TryRead<DateTime>(key, out _));
+        }
+
+        [Fact]
+        public void TryRead_String_As_Numeric_Should_Succeed()
+        {
+            const string key = "test";
+            const int val = 10;
+            var tc = CreateTwinCollectionReader(key, val.ToString(CultureInfo.InvariantCulture));
+            Assert.True(tc.TryRead<int>(key, out var n));
+            Assert.Equal(val, n);
+        }
+
+        [Fact]
+        public void TryRead_Can_Convert_Numbers_To_Enum()
+        {
+            const string key = "test";
+            var tc = CreateTwinCollectionReader(key, (int)DataRateIndex.DR8);
+            Assert.True(tc.TryRead<DataRateIndex>(key, out var dr));
+            Assert.Equal(DataRateIndex.DR8, dr);
+        }
+
+        [Fact]
+        public void Undefined_Enum_Values_Are_Not_Parsed()
+        {
+            const string key = "test";
+            var tc = CreateTwinCollectionReader(key, 100);
+            Assert.False(tc.TryRead<DataRateIndex>(key, out _));
```

Comment:
[nit] should we also assert the value of the discarded result here?

Reply:
the result is default when the method returns false. That's another test.